src/server.lisp

```diff
@@ -0,0 +1,184 @@
+                                                :method-name (|RPCRequest-method| request))))
+                        (setf result 
+                              (if timeout
+                                  (bt:with-timeout (timeout)
+                                    (apply f params-as-plist))
+                                  (apply f params-as-plist)))
+                        
+                        (setf reply (make-instance '|RPCReply|
+                                                   :|id| (|RPCRequest-id| request)
+                                                   :|result| result))))
+                    (format-log logging-stream "Finishing request: ~a" (|RPCRequest-method| request)))
+                
+                ;; this is where errors go where we can reply to the client
+                (unknown-RPC-method (c)
+                  (declare (ignore c))
+                  (format-log logging-stream "Error: method ~a unknown" (|RPCRequest-method| request))
+                  (setf reply (make-instance '|RPCError|
+                                             :|id| (|RPCRequest-id| request)
+                                             :|error| (format nil "Method named \"~a\" is unknown."
+                                                              (|RPCRequest-method| request)))))
+                (sb-ext:timeout (c)
+                  (declare (ignore c))
+                  (format-log logging-stream "Timed out on request ~a" (|RPCRequest-method| request))
+                  (setf reply (make-instance '|RPCError|
+                                             :|id| (|RPCRequest-id| request)
+                                             :|error| (format nil "Execution timed out.  Note: time limit: ~a seconds." timeout))))
+                (simple-error (c)
+                  (format-log logging-stream "Threw generic error during RPC call:~%~a" c)
+                  (setf reply (make-instance '|RPCError|
+                                             :|id| (|RPCRequest-id| request)
+                                             :|error| (format nil "~a" c)))))
+              
+              ;; send the client response, whether success or failure
+              (%push-raw-request receiver identity empty-frame (serialize reply))))
+        
+        ;; this is where errors go where we can't even reply to the client
+        (simple-error (c)
+          (format-log logging-stream "Threw generic error before RPC call:~%~a" c))))))
+
+(defun start-server (&key
+                       dispatch-table
+                       (listen-addresses (list "tcp://*:5555"))
+                       (thread-count 5)
+                       (logging-stream (make-broadcast-stream))
+                       timeout)
+  "Main loop of an RPCQ server. Spawns THREAD-COUNT many threads which service LISTEN-ADDRESSES by replying to RPCQ requests that match entries in DISPATCH-TABLE. Kills any RPC calls taking longer than TIMEOUT seconds."
```

Comment:
Can you describe the types more? The descriptions can be on the next (few) line(s) and not cramped into the middle of this.

We should also check the types at the entrypoint.